### PR TITLE
Adds a toggle so ghosts can hear radio chatter more easily.

### PR DESCRIPTION
**Pull request by theOperand**
ITS HABBEDING

and thus we run out of space on our toggles variable.

### CODE REVIEW

**Comment by Steelpoint** [comment]
:+1:

**Comment by RemieRichards** [comment]
@Miauw62 can you comment on the toggle variable stating that it's full so we don't get people trying to add more to it?

**Comment by GunHog** [comment]
YES! <b>YES, PLEASE!</b> I am so tired of seating my ghost on the AI/Syndicate intercom in order to hear the radio!

**Comment by AlexCrimson** [comment]
Oh my god. As someone who has probably observed more rounds than he has played... i love you.

**Comment by Ikarrus** [comment]
The Administration gives this a :+1:

**Comment by Menshin** [comment]
> and thus we run out of space on our toggles variable.

Byond variables are 16 bits ?

**Comment by RemieRichards** [comment]
@Menshin Yes.

**Comment by Dorsisdwarf** [comment]
Ghostradio and GhostPDA should be the same toggle.

I can't think of any circumstance you'd want to listen to radio and not PDA, and it saves a slot.

**Comment by Incoming5643** [comment]
I'm pretty sure you can get around interger size limits for bitflags in byond by having values iterate DOWN from 1 instead of up:

#DEFINE SOME_FLAG 1
#DEFINE ANOTHER_FLAG 0.5
#DEFINE YET_MORE_FLAG 0.25
.
.
.

BUT if you do that the whole toggles variable would have to be resanitized and everyone would have there setting for it reset unless every savefile was specially converted beforehand

**Comment by RemieRichards** [comment]
@Incoming5643 I really doubt that works.
Logic would assume it would still cap out at 16 iterations, be it for a technical reason or just for consistency.

**Comment by Incoming5643** [comment]
I've seen it used like that in the code in several places, but I'm not sure if it actually allows for more variables or is just a taste thing. Hey I don't see why you couldn't have it iterating both ways at once, that's a surefire way to squeeze a few more flags in there, even if it makes it look a little dumb

**Comment by AlexCrimson** [comment]
Also need an AntagHUD. Just sayin'

**Comment by RemieRichards** [comment]
@Incoming5643 I have never seen 1, 0.5, 0.25 anywhere in the code so please share.

I've seen stupid bitshifting on constants for antags, and whatever the fuck list2text() is but besides that, not much fuckery.

**Comment by Incoming5643** [comment]
I can't find it either now, but I could have sworn at one point I saw someone use it like that :sweat_smile:

**Comment by RemieRichards** [comment]
Hmm, a likely story! Hehe, I'll give you the benefit of the doubt.

**Comment by Menshin** [comment]
There's no way this "down iteration" thing would work: iirc, Byond numbers are floats, even integers except when used along bitwise operators, where they are typecasted as int.

0.5 is not an integer, it's representation is whatever Byond is using for float (pretty sure it's standard internally , so bitsign|mantissa|exponant). Using bitwise operators with these is obviously not a good idea.

**Comment by Alek2ander** [comment]
65536 wouldn't fit in a 16 bit var...

File: code/__DEFINES/preferences.dm [hunk 1]
```diff
@@ -16,8 +16,9 @@
 #define CHAT_GHOSTWHISPER 8192
 #define MIDROUND_ANTAG	16384
 #define CHAT_GHOSTPDA	32768
+#define CHAT_GHOSTRADIO 65536
 
-#define TOGGLES_DEFAULT (SOUND_ADMINHELP|SOUND_MIDI|SOUND_AMBIENCE|SOUND_LOBBY|CHAT_OOC|CHAT_DEAD|CHAT_GHOSTEARS|CHAT_GHOSTSIGHT|CHAT_PRAYER|CHAT_RADIO|MEMBER_PUBLIC|CHAT_PULLR|INTENT_STYLE|CHAT_GHOSTWHISPER|MIDROUND_ANTAG|CHAT_GHOSTPDA)
+#define TOGGLES_DEFAULT (SOUND_ADMINHELP|SOUND_MIDI|SOUND_AMBIENCE|SOUND_LOBBY|CHAT_OOC|CHAT_DEAD|CHAT_GHOSTEARS|CHAT_GHOSTSIGHT|CHAT_PRAYER|CHAT_RADIO|MEMBER_PUBLIC|CHAT_PULLR|INTENT_STYLE|CHAT_GHOSTWHISPER|MIDROUND_ANTAG|CHAT_GHOSTPDA|CHAT_GHOSTRADIO)
```
Comment (RemieRichards) [inline code/__DEFINES/preferences.dm]: Wait there are 17 toggles in this now, isn't the cap 16?

Comment (Miauw62) [inline code/__DEFINES/preferences.dm]: wait fuck i thought it went up to 17

Comment (RemieRichards) [inline code/__DEFINES/preferences.dm]: Well it's *16* bit, I'm pretty sure it caps at 16.

**Comment by Miauw62** [comment]
moving all these chat toggles into their own var atm, give me a sec.

**Comment by AlexCrimson** [comment]
Make each one half a bit, then you will have 32!

**Comment by Miauw62** [comment]
:^]

**Comment by Miauw62** [comment]
there, i changed the variable to chat_toggles. also added a missing preferences button.

File: code/modules/client/preferences.dm [hunk 2]
```diff
@@ -221,11 +222,12 @@ datum/preferences
...
-				dat += "<b>Ghost whispers:</b> <a href='?_src_=prefs;preference=ghost_whispers'>[(toggles & CHAT_GHOSTWHISPER) ? "Nearest Creatures" : "All Speech"]</a><br>"
-				dat += "<b>Ghost pda:</b> <a href='?_src=prefs;preference=ghost_pda'>[(toggles & CHAT_GHOSTPDA) ? "Nearest Creatures" : "All Messages"]</a><br>"
-				dat += "<b>Pull requests:</b> <a href='?_src_=prefs;preference=pull_requests'>[(toggles & CHAT_PULLR) ? "Yes" : "No"]</a><br>"
+				dat += "<b>Ghost ears:</b> <a href='?_src_=prefs;preference=ghost_ears'>[(chat_toggles & CHAT_GHOSTEARS) ? "Nearest Creatures" : "All Speech"]</a><br>"
+				dat += "<b>Ghost sight:</b> <a href='?_src_=prefs;preference=ghost_sight'>[(chat_toggles & CHAT_GHOSTSIGHT) ? "Nearest Creatures" : "All Emotes"]</a><br>"
+				dat += "<b>Ghost whispers:</b> <a href='?_src_=prefs;preference=ghost_whispers'>[(chat_toggles & CHAT_GHOSTWHISPER) ? "Nearest Creatures" : "All Speech"]</a><br>"
+				dat += "<b>Ghost radio:</b> <a href='?_src=prefs;preference=ghost_radio'>[(chat_toggles & CHAT_GHOSTRADIO) ? "Yes" : "No"]</a><br>"
```
Comment (Miauw62) [inline code/modules/client/preferences.dm]: not entirely sure how correct this line is.

**Comment by Miauw62** [comment]
a very quick test using the osay verb tells me that this works.

**Comment by Miauw62** [comment]
>tfw this still isnt merged

**Comment by GunHog** [comment]
Rewrite the say() fonts while you wait :)

**Pull request merged**